### ISSUE TITLE
PHP5-FPM fails to detect ports in use, fails to startup.

### ISSUE THREAD

**Issue by chris001** [webmin/webmin#518]
Service `php5-fpm` fails to start.  Port 8000 is already in use by another http server process (not apache), in fact the other server process running on port 8000 is managed by a standard webmin module, and post 8000 is set in its settings, and the port number is visible on the webmin servers web interface.
```
Mar 18 01:01:13 server1 php5-fpm[8840]: [18-Mar-2017 01:01:13] ERROR: unable to bind listening socket for address '8000': Address already in use (98)
Mar 18 01:01:13 server1 php5-fpm[8840]: [18-Mar-2017 01:01:13] ERROR: FPM initialization failed
Mar 18 01:01:13 server1 systemd[1]: php5-fpm.service: main process exited, code=exited, status=78/n/a
Mar 18 01:01:13 server1 systemd[1]: Failed to start The PHP FastCGI Process Manager.
Mar 18 01:01:13 server1 systemd[1]: Unit php5-fpm.service entered failed state.
Mar 18 01:03:30 server1 systemd[1]: Stopped The PHP FastCGI Process Manager.
```

**Comment by jcameron** [comment]
Virtualmin doesn't actually configure FPM to use port 8000 though - that's the port that the distribution package uses by default. 

What else is on port 8000?

**Comment by chris001** [comment]
@jcameron Port 8000 is configured for use by the icecast2 mp3 streaming server.

**Comment by jcameron** [comment]
What I'd recommend is to just edit `/etc/php*/fpm/pool.d/www.conf` and change the port number from 8000 to something else, then restart the FPM server. That will resolve the port clash.

**Comment by chris001** [comment]
Sure about that? There's no `8000` in `/etc/php*/fpm/pool.d/www.conf`.
This is (appears to be) the relevant section of the file:
```
; The address on which to accept FastCGI requests.
; Valid syntaxes are:
;   'ip.add.re.ss:port'    - to listen on a TCP socket to a specific IPv4 address on
;                            a specific port;
;   '[ip:6:addr:ess]:port' - to listen on a TCP socket to a specific IPv6 address on
;                            a specific port;
;   'port'                 - to listen on a TCP socket to all IPv4 addresses on a
;                            specific port;
;   '[::]:port'            - to listen on a TCP socket to all addresses
;                            (IPv6 and IPv4-mapped) on a specific port;
;   '/path/to/unix/socket' - to listen on a unix socket.
; Note: This value is mandatory.
listen = /var/run/php5-fpm.sock
```

**Comment by virtualmin** [comment]
Ok .. does the number 8000 appear anywhere under `/etc/php*/fpm` ?

**Comment by chris001** [comment]
Unfortunately 8000 doesn't appear anywhere under `/etc/php*/fpm` !
Here's what that directory contains:
```
drwxr-xr-x 2 root root  4096 Oct 26 17:21 conf.d
-rw-r--r-- 1 root root  4551 Jul 26  2016 php-fpm.conf
-rw-r--r-- 1 root root 72664 Dec 13 11:23 php.ini
drwxr-xr-x 2 root root  4096 Mar 18 03:02 pool.d
/etc/php5/fpm#
```
Then inside `pool.d` is just one file, `www.conf`.  `conf.d` contains just symlinks to php module `.ini` files.

**Comment by jcameron** [comment]
Now I'm confused - if port 8000 doesn't appear anywhere in the FPM config files, why would it be clashing with another server on port 8000 ?

**Comment by chris001** [comment]
@jcameron Good question. It must be pulling the 8000 from another file located elsewhere, or possibly the apache proxy module or php5-fpm is generating that port number thru default code in the face of a lack of a config file default.
I've got more details on this bug for you.
The fpm ports seem to "stick" to the virtual server where the port was originally assigned.  For example, the first virtual server ever configured as php-fpm on this system, several weeks ago, call it sever A, was assigned 8000, then reverted to fcgid mode.  Today, turned a different virtual server, call it server B, to php5-fpm mode, and it worked, since it was assigned port 8001.  Go back and turn server A to php5-fpm, and it fails, since it wants to try port 8000 which is already in use by another separate web service.  Try to put virtual server C into php-fpm mode, and it would succeed since the system assigns port 8002, yet it fails because server A is still in php-fpm mode and carelessly trying to use port 8000.
```
Changing PHP execution mode to FPM (run as virtual server owner) ..
.. done

Changing maximum PHP script run time to 30 seconds ..
.. done

Restarting PHP-FPM server ..
.. failed : Starting php5-fpm (via systemctl): php5-fpm.serviceJob for php5-fpm.service failed. See 'systemctl status php5-fpm.service' and 'journalctl -xn' for details. failed!

Applying web server configuration ..
.. done
```

**Comment by jcameron** [comment]
Right, the ports are re-used where possible.

Is the problem actually what when you re-enable FPM for domain A, it picks port 8000 even though it is in use by another web app?

**Comment by chris001** [comment]
Right, @jcameron  that's exactly it, when you re-enable FPM for domain A, it picks port 8000, which is in use by another web app, and has been in use by the other web app, `icecast2`, since way before FPM first became available in Virtualmin.  FPM has consistently failed for this domain A, because it keeps trying, every single time you go to enable FPM on the domain A, it tries to use port 8000, which has always been in use, and that has made FPM always fail to restart.

It'd be real nice if virtualmin or fpm itself could autodetect when a port it's considering for take over as an FPM port, is actually in use, and when it is, skip up to the next port, and try again.  This would make enabling fpm fail proof.  You could enable fpm for all servers, and it'd work, without that dreaded vague error, ```
Job for php5-fpm.service failed. See 'systemctl status php5-fpm.service' and 'journalctl -xn' for details.
```

**Comment by chris001** [comment]
The port 8000 on localhost appears to be firewalled/blocked!  Yet 8000 does respond on the network interfaces with the routable public IP address however.  It's an odd corner case, eh?
```
server1:~# telnet localhost 8000
Trying ::1...
Trying 127.0.0.1...
telnet: Unable to connect to remote host: Connection refused
```

**Comment by chris001** [comment]
Probably `netstat` is a more reliable indicator of whether something is listening on a port, since it ignores how the firewall changes the behavior depending on who is connecting from where?
```
server1:~# netstat -lp | grep 8000
tcp        0      0 server1.edomain:8000 *:*                     LISTEN      788/icecast2
unix  2      [ ACC ]     STREAM     LISTENING     4279180003 1805/master         private/discard
unix  2      [ ACC ]     STREAM     LISTENING     4279180007 1805/master         private/local
server1:~# netstat -lp | grep 8080
server1:~# netstat -lp | grep http
tcp        0      0 localhost.loca:http-alt *:*                     LISTEN      10779/apache2
tcp        0      0 *:http                  *:*                     LISTEN      31482/varnishd
tcp        0      0 *:https                 *:*                     LISTEN      2545/nginx: worker
tcp6       0      0 [::]:http               [::]:*                  LISTEN      31482/varnishd
tcp6       0      0 [::]:https              [::]:*                  LISTEN      2545/nginx: worker
```

**Comment by jcameron** [comment]
Right, I've added code to also use `netstat` to find ports in use.

**Comment by MurzNN** [comment]
After upgrade to last versions:
```
   webmin (1.900 => 1.921)
   webmin-virtual-server (6.06.gpl-2 => 6.07.gpl)
```
all ports are changed ([more info here](https://virtualmin.com/comment/815162#comment-815162)) - this is not so critical (bur very bad), other problem is that one of virtualserver got already busy port 8080 and php-fpm fail to restart with error:
```
ERROR: unable to bind listening socket for address '8080': Address already in use (98)
```
This port is used by other process:
```
$ netstat -lpnN | grep 8080
tcp6       0      0 :::8080                 :::*                    LISTEN      10005/gulp          
```

**Comment by sz00gun** [comment]
Any fix for that?

**Comment by jcameron** [comment]
Was it Apache that was trying to use port 8080 and failing to start, or an FPM server? How did you configure this port in Virtualmin?

**Comment by sz00gun** [comment]
I don't know, I didn't configure anything. I just use Virtualmin out of box. After last Virtualmin update last week all went down. 

All goes down even after doing: check configuration in Virtualmin. 

Also fpm goes down after  check server security in CFS. 

A lot information from members here https://www.virtualmin.com/node/66694

**Comment by alex-pimperton** [comment]
We've also got the same issue across all our FPM virtualmin servers, we have no clashing ports but as soon as a check config is done all the ports change in FPM but are not updated in NGINX and the websites all fall over. Flipping all over to FCGID and back to FPM fixes it in the short term.

**Comment by sz00gun** [comment]
You are not alone mate. I'm going to make a presentation of bugs in webmin today later...

Hopefully that catch developers.

**Comment by jcameron** [comment]
The ports being changed un-necessarily is a bug - that feature was only supposed to fix clashes due to multiple PHP-FPM packages being installed with conflicting default ports. The next release will fix it.

**Comment by MurzNN** [comment]
I fixed port changes manually on after upgrading to webmin 1.921, but after next upgrade webmin (1.921 => 1.930) - port changes again! :angry:

**Comment by sz00gun** [comment]
I mentioned about presentation yesterday.

It's here:  https://www.virtualmin.com/node/66956

@jcameron please have a look too.

**Comment by jcameron** [comment]
@MurzNN - this requires an update to the Virtualmin package, which is separate from Webmin

**Comment by MurzNN** [comment]
@jcameron Does Virtualmin package update already available with fixing port changing on each configuration re-check?

**Comment by jcameron** [comment]
We haven't released that update yet
